samza-api/src/main/java/org/apache/samza/context/ApplicationDefinedContainerContext.java

```diff
@@ -0,0 +1,22 @@
+package org.apache.samza.context;
+
+/**
+ * An application should implement this to contain any runtime objects required to initialize the processing logic.
+ * <p>
+ * A single instance of this will be created in each container, so it will be shared between all of the tasks on a
+ * single container.
+ * <p>
+ * If it is necessary to have a separate instance per task, then use {@link ApplicationDefinedTaskContext} instead.
+ * An {@link ApplicationDefinedContainerContextFactory} needs to be implemented to build instances of this context.
+ */
+public interface ApplicationDefinedContainerContext {
```

Comment:
One additional followup question with respect to standalone. 

Are we planning to invoke the `ApplicationDefinedContainerContext` start/stop once for a standalone stream processor or everytime before/after a rebalancing phase(when we kill/create a SamzaContainer as a part of rebalancing phase).

1. Existing `SamzaContainer` setup/teardown is expensive and setting the right coordination timeouts is a known problem for the standalone users. Just wondering if this will compound that problem even further.
2. In standalone, let's say if a user wants to instantiate some heavy-weight, expensive objects only once(using samza context) per java process, then what is the recommended approach after this change?

Reply:
The plan was to invoke start/stop whenever we create/kill a SamzaContainer.
You are right that this could possibly impact standalone during a rebalance. In general, I don't think start and stop takes a long time, but it would take some time, and there may be cases where it takes a long time. It is a good point that this would also change the existing behavior of when some objects are created in standalone. The difficulty is that some of the app-defined context does actually need to be recreated (e.g. what if the context depends on the container-partition mapping?). Do you think it is worth it to add additional complexity to differentiate between context that needs to be recreated and context that does not need to be recreated?

Reply:
Had some discussions in person about this:
1) Since YARN deployments still do not use StreamProcessor, it will be difficult to achieve consistency between YARN and standalone with a processor-level context.
2) The concrete use cases we came up with for application-defined processor context can currently go in the container-level context (it seems ok to recreate them).
3) Use cases like job coordinator lifecycle are not used by applications and they do not depend on application code, so it is currently unnecessary to add something to the public API for that.
4) It is still possible to extend this API to have a processor-level context in the future.